src/azure-cli/azure/cli/command_modules/storage/operations/azcopy.py

```diff
@@ -11,82 +11,75 @@
-                 follow_symlinks=None):
-    def get_url_with_sas(source, account_name, container, blob, share, file_path, local_path):
-        import re
-        import os
-        from azure.cli.command_modules.storage._validators import _query_account_key
-        from azure.cli.command_modules.storage.azcopy.util import _generate_sas_token
-        storage_endpoint = cmd.cli_ctx.cloud.suffixes.storage_endpoint
-        if source is not None:
-            if "?" in source:  # sas token exists
-                return source
-            storage_pattern = re.compile(r'https://(.*?)\.(blob|dfs|file).%s' % storage_endpoint)
-            result = re.findall(storage_pattern, source)
-            if result:   # source is URL
-                storage_info = result[0]
-                account_name = storage_info[0]
-                if storage_info[1] in ['blob', 'dfs']:
-                    service = 'blob'
-                elif storage_info[1] in ['file']:
-                    service = 'file'
-                else:
-                    raise ValueError('Not supported service type')
-                account_key = _query_account_key(cmd.cli_ctx, account_name)
-            else:   # source is path
-                return source
-        elif account_name:
-            account_key = _query_account_key(cmd.cli_ctx, account_name)
-            if container:
-                client = blob_data_service_factory(cmd.cli_ctx, {'account_name': account_name})
-                if blob is None:
-                    blob = ''
-                source = client.make_blob_url(container, blob)
+def get_url_with_sas(cmd, url, account_name, container, blob, share, file_path, local_path, **kwargs):
+    import re
+    import os
+    from azure.cli.command_modules.storage._validators import _query_account_key
+    from azure.cli.command_modules.storage.azcopy.util import _generate_sas_token
+    storage_endpoint = cmd.cli_ctx.cloud.suffixes.storage_endpoint
+    account_key = kwargs.pop('account_key', None)
+    connection_string = kwargs.pop('connection_string', None)
+    sas_token = kwargs.pop('sas_token', None)
+    if connection_string:
+        from azure.cli.core.commands.validators import validate_key_value_pairs
+        conn_dict = validate_key_value_pairs(connection_string)
+        account_name = conn_dict.get('AccountName')
+        account_key = conn_dict.get('AccountKey')
+        sas_token = conn_dict.get('SharedAccessSignature')
+
+    if url is not None:
+        if "?" in url:  # sas token exists
+            return url
+        # validate source is uri or local path
+        storage_pattern = re.compile(r'https://(.*?)\.(blob|dfs|file).%s' % storage_endpoint)
+        result = re.findall(storage_pattern, url)
+        if result:  # source is URL
+            storage_info = result[0]
+            account_name = storage_info[0]
+            if storage_info[1] in ['blob', 'dfs']:
```

Comment:
Could we consider reuse some common logics from https://github.com/Azure/azure-cli/blob/cfa31a7439e3d7dd4ff38fe03afd8a753a9cb19b/src/azure-cli/azure/cli/command_modules/storage/_validators.py#L1347